### PR TITLE
Collection literals: create methods

### PR DESCRIPTION
**Pull request by cston**
Update section for custom create methods based on recent discussions.

### CODE REVIEW

File: proposals/collection-literals.md
Comment (CyrusNajmabadi) [inline proposals/collection-literals.md:188]: seems like it would be fine to be on an interface type.  indeed, you could imagine adding this to [`IImutableSet<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.iimmutableset-1?view=net-7.0)

File: proposals/collection-literals.md
Comment (RikkiGibson) [inline proposals/collection-literals.md:203]: I would rephrase this as "if the parameter does not escape the Create method". Alternatively we could define a concept of "implicitly scoped" and use it here. All I really want to avoid terminology-wise is having to say in several places "if the method does not return ref struct or have a writable ref parameter of ref struct type etc..."

I'm OK with this limitation as it applies to user-defined ref struct types. I somewhat feel that if you are creating a ref struct which is simply wrapping a span, what are you really accomplishing. I don't know whether that's a real use case.

If such ref structs were a real use case, though, my feeling would be that this is unnecessarily limiting. Why shouldn't my ref-struct be able to hold on to stackalloc'd memory just like a Span would. In that case I would prefer a rule that if a collection-expr is converted to ref-struct, its safe-to-escape is the current local scope. It's a simple rule which remains "intact" even through incidental refactorings. If you don't want that behavior, convert the CL to non-ref-struct type first. This is essentially what we suggested for CLs converted directly to Span, I think.

**Review by RikkiGibson — APPROVED** [review]

File: proposals/collection-literals.md
Comment (333fred) [inline proposals/collection-literals.md:207]: This is striking me as `infoof`. Do we _really_ need to be able to specify member names here, and not just have a well-recognized pattern that the type indicated must follow?

Comment (CyrusNajmabadi) [reply proposals/collection-literals.md:211]: @333fred i believe we need this.  But i'll let @stephentoub weigh in.  I think the core issue is that the BCL expects (or already has) differently named methods in a few places, and they'd want these attributes to point at them (versus adding a new duplicative method).  Because of that, we need a type+methodname pair to clearly identify things.  

Note: i believe this is going through Runtime review soon (or maybe already has).  So i'm ok deferring to what they want.  ON our side we just need something to go find the appropriate method :)

Comment (stephentoub) [reply proposals/collection-literals.md:211]: @333fred, are you suggesting that the compiler would look at all members of the specified type for one with the right shape and use whatever it finds?

File: proposals/collection-literals.md
Comment (jaredpar) [inline proposals/collection-literals.md:207]:
```suggestion
If the parameter does not escape the *create method* (that is, if the parameter is `scoped`, the return type is not a `ref struct` and there are no `out / ref` of type `ref struct`), the compiler will assign a lifetime of _current method_ to the argument. As such it *may* allocate the storage for the `ReadOnlySpan<T0>` on the stack rather than the heap. 
```

File: proposals/collection-literals.md
Comment (jaredpar) [inline proposals/collection-literals.md:216]:
```suggestion
    public static ImmutableArray<T> Create<T>(scoped ReadOnlySpan<T> items) { ... }
```

Ideally this parameter is `scoped`

File: proposals/collection-literals.md
Comment (jaredpar) [reply proposals/collection-literals.md:207]: Do not see any discussion about the lifetime of a collection literal that is used in a local declaration.

File: proposals/collection-literals.md
Comment (333fred) [reply proposals/collection-literals.md:211]: > @333fred, are you suggesting that the compiler would look at all members of the specified type for one with the right shape (accepting a `ReadOnlySpan<T>`, returning a `CollectionType`, name doesn't matter) and use whatever it finds?

Yes, but I would expect the name to be part of the pattern it looks for.

> I think the core issue is that the BCL expects (or already has) differently named methods in a few places, and they'd want these attributes to point at them (versus adding a new duplicative method).

We've already done this multiple times for language features. For example, `List<T>` added the `Slice` method which just defers to `GetRange`, to work with the slicing feature. I also wonder, if we have some small list of canonical methods, can we not just codify them now? Is there a list of the existing method names that we would want to consider for the pattern?

File: proposals/collection-literals.md
Comment (cston) [reply proposals/collection-literals.md:207]: I didn't mention `out` or `ref` parameters here because the *create method* pattern is specifically `static CollectionType<T> Create<T>(ReadOnlySpan<T> items)`, with a single parameter passed by value.

File: proposals/collection-literals.md
Comment (stephentoub) [reply proposals/collection-literals.md:211]: > Yes, but I would expect the name to be part of the pattern it looks for.

That'll work for the specific uses planned for the core libraries for .NET 8, which are all in immutable collections.  It won't work when we add ctors to types like `HashSet<T>` and expect those to be usable.  It won't work when we add additional patterns and want multiple named methods on the same type, e.g. a CollectionsMarshal.CreateList and a CollectionsMarshal.CreateQueue.  And it won't work for other collections that already have an appropriately-shaped factory method with a different name, forcing the developer to add another API for it that does the same thing.

I think we should continue having the name be specified.  It's more flexible and specific and it doesn't hurt anything.

> For example, List<T> added the Slice method which just defers to GetRange, to work with the slicing feature

And we didn't like doing so.  We only did because `List<T>` is such a prominent type and it was the least friction-free way to unblock that use.  I do not see us, for example, adding a static `List` class with a static `Create` method.

File: proposals/collection-literals.md
Comment (jaredpar) [reply proposals/collection-literals.md:207]: True but shouldn't we discuss the more general policy here? Specifically that collection literals when passed to a method have such and such rules. The rules for custom create methods aren't special here, they just fall out from those broader rules.

Comment (jaredpar) [reply proposals/collection-literals.md:207]: Basically there is no diff between 

```csharp

ImmutableArray<int> array1 = { 1, 2, 3 };
ImmutableArray<int> array2 = MyCreate({1, 2, 3 });
ImmutableArray<int> MyCreate(scoped ReadOnlySpan<int> span) => ...
```

File: proposals/collection-literals.md
Comment (333fred) [reply proposals/collection-literals.md:211]: Ctors seem like something we codify. The multiple methods also seems like something we can also codify, with a rule involving the type name. Every time we've looked at something like `methodof` or `infoof` in C#, we've arrived at all sorts of issues that need to be resolved. I'm not innately opposed to it, but as this as historically been a difficult problem for us to address, we might want to consider a more hardcoded list.

File: proposals/collection-literals.md
Comment (cston) [reply proposals/collection-literals.md:127]: > Is there a need to lock the create method in to taking ReadOnlySpan?

Restricting the pattern to `ReadOnlySpan<T>` is likely to simplify testing, especially in cases where there are overloads. We can expand this to other types if there are scenarios.

File: proposals/collection-literals.md
Comment (cston) [reply proposals/collection-literals.md:207]: The comment here was to clarify the requirements for this particular case. We could perhaps remove this comment when we add a more general statement about lifetime in a subsequent PR.

File: proposals/collection-literals.md
Comment (cston) [reply proposals/collection-literals.md:216]: I had this as `scoped` initially but removed the modifier from this example since it's redundant.

**Pull request merged**
